plugin/db/mysql/restore.go

```diff
@@ -499,6 +504,43 @@ func (driver *Driver) downloadBinlogFilesOnServer(ctx context.Context, binlogFil
 
 // FetchAllBinlogFiles downloads all binlog files on server to `binlogDir`.
 func (driver *Driver) FetchAllBinlogFiles(ctx context.Context) error {
+	// Ensure that there's at most one ongoing downloading process for the current MySQL instance.
```

Comment:
What's the consequence if not?

Reply:
There might be concurrent writing to the same binlog file and data corruption occurs.

Reply:
Second thought: All binlog file operations (download/replay) should be serialized.
I'll try to use a mutex to do this.

Reply:
What will happen if we don't use mutex? I thought we download to a temp file and rename it later?

Reply:
We are not using the temp file mechanism, because it'll introduce more potential IO errors and concurrent issues.
I prefer to do the locking explicitly in the code, which is more reliable.

Reply:
I prefer to temp file mechanism if the rename operation is atomic.

Reply:
The concern is that the filesystem operations are not necessarily atomic, at least for Linux. And different fs may behave differently. @h3n4l

Reply:
> The concern is that the filesystem operations are not necessarily atomic, at least for Linux. And different fs may behave differently. @h3n4l

That's what worries me too...

Reply:
Per offline discussion and experiments, the concurrent write processes will produce a consistent file with the flag `O_WRONLY`. It's unnecessary to use locks to serialize the operations.